presto-main/src/main/java/com/facebook/presto/server/AsyncPageTransportServlet.java

```diff
@@ -0,0 +1,208 @@
+import static javax.servlet.http.HttpServletResponse.SC_BAD_REQUEST;
+import static javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
+import static javax.servlet.http.HttpServletResponse.SC_NO_CONTENT;
+
+public class AsyncPageTransportServlet
+        extends HttpServlet
+{
+    private static final Logger log = Logger.get(AsyncPageTransportServlet.class);
+
+    private final Duration pageTransportTimeout;
+    private final TaskManager taskManager;
+    private final Executor responseExecutor;
+    private final ScheduledExecutorService timeoutExecutor;
+
+    @Inject
+    public AsyncPageTransportServlet(
+            TaskManager taskManager,
+            ExchangeClientConfig exchangeClientConfig,
+            @ForAsyncRpc BoundedExecutor responseExecutor,
+            @ForAsyncRpc ScheduledExecutorService timeoutExecutor)
+    {
+        this.taskManager = requireNonNull(taskManager, "taskManager is null");
+        this.pageTransportTimeout = requireNonNull(exchangeClientConfig.getAsyncPageTransportTimeout(), "asyncPageTransportTimeout is null");
+        this.responseExecutor = requireNonNull(responseExecutor, "responseExecutor is null");
+        this.timeoutExecutor = requireNonNull(timeoutExecutor, "timeoutExecutor is null");
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response)
+            throws IOException
+    {
+        String requestURI = request.getRequestURI();
+        // Example:  /v1/task/async/{taskId}/results/{bufferId}/{token}
+        List<String> requestURIParts = Arrays.asList(requestURI.split("/"));
+
+        TaskId taskId = TaskId.valueOf(requestURIParts.get(4));
+        OutputBufferId bufferId = OutputBufferId.fromString(requestURIParts.get(6));
+        long token = parseLong(requestURIParts.get(7));
+        DataSize maxSize = DataSize.valueOf(request.getHeader(PRESTO_MAX_SIZE));
+
+        if (requestURIParts.size() != 8) {
+            response.sendError(SC_BAD_REQUEST, format("Unexpected URI for task result request in async mode: %s", requestURI));
+            return;
+        }
+
+        AsyncContext asyncContext = request.startAsync(request, response);
+        // wait time to get results
+        Duration waitTime = randomizeWaitTime(DEFAULT_MAX_WAIT_TIME);
+        asyncContext.setTimeout(waitTime.toMillis() + pageTransportTimeout.toMillis());
+
+        asyncContext.addListener(new AsyncListener()
+        {
+            public void onComplete(AsyncEvent event)
+            {
+            }
+
+            public void onError(AsyncEvent event)
+                    throws IOException
+            {
+                String errorMessage = format("Server error to process task result request %s : %s, ", requestURI, event.getThrowable().getMessage());
+                log.error(errorMessage);
+                response.setStatus(SC_INTERNAL_SERVER_ERROR);
+                response.setHeader(CONTENT_TYPE, PRESTO_PAGES);
```

Comment:
Since you are sending an error any way -there's no need in setting all these headers.

Reply:
Responses with error code other than 200 are still retried: 

https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/HttpRpcShuffleClient.java#L145 - throws an exception

https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/PageBufferClient.java#L386  - retries without checking the headers